external/storm-hdfs/src/main/java/org/apache/storm/hdfs/blobstore/HdfsBlobStore.java

```diff
@@ -0,0 +1,379 @@
+    BlobStoreAclHandler.validateSettableACLs(key, meta.get_acl());
+    SettableBlobMeta orig = getStoredBlobMeta(key);
+    _aclHandler.hasPermissions(orig.get_acl(), ADMIN, who, key);
+    BlobStoreFileOutputStream mOut = null;
+    try {
+      BlobStoreFile hdfsFile = _hbs.write(META_PREFIX + key, false);
+      hdfsFile.setMetadata(meta);
+      mOut = new BlobStoreFileOutputStream(hdfsFile);
+      mOut.write(Utils.thriftSerialize(meta));
+      mOut.close();
+      mOut = null;
+    } catch (IOException e) {
+      throw new RuntimeException(e);
+    } finally {
+      if (mOut != null) {
+        try {
+          mOut.cancel();
+        } catch (IOException e) {
+          //Ignored
+        }
+      }
+    }
+  }
+
+  @Override
+  public void deleteBlob(String key, Subject who)
+      throws AuthorizationException, KeyNotFoundException {
+    who = checkAndGetSubject(who);
+    validateKey(key);
+    SettableBlobMeta meta = getStoredBlobMeta(key);
+    _aclHandler.hasPermissions(meta.get_acl(), WRITE, who, key);
+    try {
+      _hbs.deleteKey(DATA_PREFIX + key);
+      _hbs.deleteKey(META_PREFIX + key);
+    } catch (IOException e) {
+      throw new RuntimeException(e);
+    }
+  }
+
+  @Override
+  public InputStreamWithMeta getBlob(String key, Subject who)
+      throws AuthorizationException, KeyNotFoundException {
+    who = checkAndGetSubject(who);
+    validateKey(key);
+    SettableBlobMeta meta = getStoredBlobMeta(key);
+    _aclHandler.hasPermissions(meta.get_acl(), READ, who, key);
+    try {
+      return new BlobStoreFileInputStream(_hbs.read(DATA_PREFIX + key));
+    } catch (IOException e) {
+      throw new RuntimeException(e);
+    }
+  }
+
+  @Override
+  public Iterator<String> listKeys() {
+    try {
+      return new KeyTranslationIterator(_hbs.listKeys(), DATA_PREFIX);
+    } catch (IOException e) {
+      throw new RuntimeException(e);
+    }
+  }
+
+  @Override
+  public void shutdown() {
+    //Empty
+  }
+
+  @Override
+  public int getBlobReplication(String key, Subject who) throws AuthorizationException, KeyNotFoundException {
+    who = checkAndGetSubject(who);
+    validateKey(key);
+    SettableBlobMeta meta = getStoredBlobMeta(key);
+    _aclHandler.hasAnyPermissions(meta.get_acl(), READ | WRITE | ADMIN, who, key);
+    try {
+      return _hbs.getBlobReplication(DATA_PREFIX + key);
+    } catch (IOException exp) {
+      throw new RuntimeException(exp);
+    }
+  }
+
+  @Override
+  public int updateBlobReplication(String key, int replication, Subject who) throws AuthorizationException, KeyNotFoundException {
+    who = checkAndGetSubject(who);
+    validateKey(key);
+    SettableBlobMeta meta = getStoredBlobMeta(key);
+    meta.set_replication_factor(replication);
+    BlobStoreFileOutputStream mOut = null;
+    _aclHandler.hasAnyPermissions(meta.get_acl(), WRITE | ADMIN, who, key);
+    try {
+      BlobStoreFile hdfsFile = _hbs.write(META_PREFIX + key, false);
```

Comment:
the actual update meta logic is very similar in createBlob, updateBlob, setBlobMeta and updateBlobReplication, I see lots of duplicate code, could you extract a helper method?